File: client_example.py
import requests
import json
import time

# Configuration
BASE_URL = "https://memory.marvn.club/api"  # Note the /api prefix
DEFAULT_TIMEOUT = 30  # seconds

class MarvinMemoryClient:
    """
    Client for interacting with Marvin's Memory API
    """
    
    def __init__(self, base_url=BASE_URL, timeout=DEFAULT_TIMEOUT):
        """
        Initialize the client
        
        Args:
            base_url: Base URL for the API
            timeout: Request timeout in seconds
        """
        self.base_url = base_url
        self.timeout = timeout
    
    def list_memories(self, page=1, limit=50, memory_type=None, min_alignment=None, tags=None):
        """
        List memories with pagination
        
        Args:
            page: Page number (starting from 1)
            limit: Maximum number of records per page
            memory_type: Filter by memory type
            min_alignment: Minimum alignment score
            tags: Filter by tags (list or comma-separated string)
            
        Returns:
            Dictionary with memories and pagination info
        """
        params = {
            "page": page,
            "limit": limit
        }
        
        if memory_type:
            params["memory_type"] = memory_type
        
        if min_alignment is not None:
            params["min_alignment"] = min_alignment
        
        if tags:
            if isinstance(tags, list):
                params["tags"] = ",".join(tags)
            else:
                params["tags"] = tags
        
        return self._make_request("GET", "/memories/", params=params)
    
    def search_memories(self, query, limit=5, memory_type=None, min_alignment=None, tags=None):
        """
        Search memories by semantic similarity
        
        Args:
            query: Search query
            limit: Maximum number of results
            memory_type: Filter by memory type
            min_alignment: Minimum alignment score
            tags: Filter by tags (list or comma-separated string)
            
        Returns:
            Dictionary with search results
        """
        params = {
            "query": query,
            "limit": limit
        }
        
        if memory_type:
            params["memory_type"] = memory_type
        
        if min_alignment is not None:
            params["min_alignment"] = min_alignment
        
        if tags:
            if isinstance(tags, list):
                params["tags"] = ",".join(tags)
            else:
                params["tags"] = tags
        
        return self._make_request("GET", "/memories/search", params=params)
    
    def create_memory(self, content, memory_type, source, tags=None):
        """
        Create a new memory
        
        Args:
            content: Memory content
            memory_type: Memory type
            source: Memory source
            tags: Memory tags (list)
            
        Returns:
            Dictionary with created memory ID
        """
        data = {
            "content": content,
            "type": memory_type,
            "source": source
        }
        
        if tags:
            data["tags"] = tags
        
        return self._make_request("POST", "/memories/", json=data)
    
    def delete_memory(self, memory_id):
        """
        Delete a memory
        
        Args:
            memory_id: Memory ID
            
        Returns:
            Dictionary with status
        """
        return self._make_request("DELETE", f"/memories/{memory_id}")
    
    def conduct_research(self, query, auto_approve=None):
        """
        Conduct research using Perplexity API
        
        Args:
            query: Research query
            auto_approve: Whether to automatically approve insights
            
        Returns:
            Dictionary with research results
        """
        data = {
            "query": query
        }
        
        if auto_approve is not None:
            data["auto_approve"] = auto_approve
        
        return self._make_request("POST", "/research/", json=data)
    
    def get_pending_research(self):
        """
        Get all pending research
        
        Returns:
            Dictionary with pending research
        """
        return self._make_request("GET", "/research/")
    
    def get_pending_research_by_id(self, query_id):
        """
        Get pending research by ID
        
        Args:
            query_id: Research query ID
            
        Returns:
            Dictionary with pending research
        """
        return self._make_request("GET", f"/research/{query_id}")
    
    def approve_insights(self, query_id, insight_indices):
        """
        Approve selected insights for storage
        
        Args:
            query_id: Research query ID
            insight_indices: List of insight indices to approve
            
        Returns:
            Dictionary with approval status
        """
        data = {
            "insight_indices": insight_indices
        }
        
        return self._make_request("POST", f"/research/{query_id}/approve", json=data)
    
    def reject_research(self, query_id):
        """
        Reject and delete pending research
        
        Args:
            query_id: Research query ID
            
        Returns:
            Dictionary with rejection status
        """
        return self._make_request("DELETE", f"/research/{query_id}")
    
    def process_tweets(self, limit=10):
        """
        Manually trigger tweet processing
        
        Args:
            limit: Maximum number of tweets to process
            
        Returns:
            Dictionary with processing results
        """
        params = {
            "limit": limit
        }
        
        return self._make_request("POST", "/tweets/process", params=params)
    
    def _make_request(self, method, endpoint, params=None, json=None, retry_count=3, retry_delay=1):
        """
        Make an HTTP request to the API with retry logic
        
        Args:
            method: HTTP method
            endpoint: API endpoint
            params: Query parameters
            json: JSON body
            retry_count: Number of retries
            retry_delay: Delay between retries in seconds
            
        Returns:
            Response JSON
        """
        url = f"{self.base_url}{endpoint}"
        
        for attempt in range(retry_count + 1):
            try:
                response = requests.request(
                    method=method,
                    url=url,
                    params=params,
                    json=json,
                    timeout=self.timeout
                )
                
                # Check if the request was successful
                response.raise_for_status()
                
                # Return the response JSON
                return response.json()
            
            except requests.exceptions.Timeout:
                if attempt < retry_count:
                    print(f"Request timed out. Retrying in {retry_delay} seconds...")
                    time.sleep(retry_delay)
                    # Increase delay for next retry
                    retry_delay *= 2
                else:
                    print("Request timed out after multiple retries.")
                    return {"error": "Request timed out"}
            
            except requests.exceptions.HTTPError as e:
                if response.status_code == 502 and attempt < retry_count:
                    print(f"Received 502 Bad Gateway. Retrying in {retry_delay} seconds...")
                    time.sleep(retry_delay)
                    # Increase delay for next retry
                    retry_delay *= 2
                else:
                    print(f"HTTP error: {e}")
                    try:
                        return response.json()
                    except:
                        return {"error": str(e), "status_code": response.status_code}
            
            except requests.exceptions.RequestException as e:
                print(f"Request error: {e}")
                return {"error": str(e)}


# Example usage
if __name__ == "__main__":
    client = MarvinMemoryClient()
    
    # Example 1: List memories with pagination
    print("\n=== Example 1: List memories with pagination ===")
    result = client.list_memories(page=1, limit=10)
    print(f"Retrieved {len(result.get('memories', []))} memories")
    print(f"Pagination info: {json.dumps(result.get('pagination', {}), indent=2)}")
    
    # Example 2: Search memories
    print("\n=== Example 2: Search memories ===")
    search_result = client.search_memories("AI consciousness", limit=3)
    print(f"Found {len(search_result.get('memories', []))} matching memories")
    
    # Example 3: Create a memory
    print("\n=== Example 3: Create a memory ===")
    create_result = client.create_memory(
        content="AI systems should be designed with ethical considerations in mind.",
        memory_type="thought",
        source="philosophical reflection",
        tags=["ethics", "AI", "philosophy"]
    )
    print(f"Created memory with ID: {create_result.get('id')}")
    
    # Example 4: Conduct research
    print("\n=== Example 4: Conduct research ===")
    research_result = client.conduct_research(
        query="What are the latest developments in AI ethics?",
        auto_approve=False
    )
    print(f"Research status: {research_result.get('status')}")
    print(f"Research ID: {research_result.get('query_id')}")

File: client_example.ts
/**
 * Marvin Memory API Client - TypeScript Version
 * 
 * This client provides methods for interacting with Marvin's Memory API.
 * It includes retry logic for handling timeouts and 502 errors.
 */

// Add Node.js type declarations for runtime environment detection
declare var process: {
  versions: {
    node: string;
  };
};
declare var require: {
  main: any;
};
declare var module: any;

// Configuration
const BASE_URL = "https://memory.marvn.club/api"; // Note the /api prefix
const DEFAULT_TIMEOUT = 30000; // 30 seconds in milliseconds

// Type definitions
export interface Memory {
  id: string;
  content: string;
  type: string;
  source: string;
  timestamp: string;
  tags: string[];
  alignment_score: number;
  similarity_score?: number;
}

export interface PaginationInfo {
  page: number;
  limit: number;
  total: number;
  pages: number;
}

export interface MemoriesResponse {
  memories: Memory[];
  pagination: PaginationInfo;
  error?: string;
}

export interface SearchResponse {
  memories: Memory[];
  error?: string;
}

export interface CreateMemoryResponse {
  id: string;
  error?: string;
}

export interface DeleteMemoryResponse {
  status: string;
  error?: string;
}

export interface ResearchResponse {
  query_id: string;
  status: string;
  insights?: any[];
  error?: string;
}

export interface PendingResearchResponse {
  pending_research: any[];
  error?: string;
}

export interface ApproveInsightsResponse {
  status: string;
  memory_ids?: string[];
  error?: string;
}

export interface ProcessTweetsResponse {
  processed: number;
  stored: number;
  error?: string;
}

export interface ListMemoriesOptions {
  page?: number;
  limit?: number;
  memoryType?: string;
  minAlignment?: number;
  tags?: string[] | string;
}

export interface SearchMemoriesOptions {
  limit?: number;
  memoryType?: string;
  minAlignment?: number;
  tags?: string[] | string;
}

export interface CreateMemoryOptions {
  content: string;
  type: string;
  source: string;
  tags?: string[];
}

export interface RequestOptions {
  params?: URLSearchParams;
  data?: any;
  retryCount?: number;
  retryDelay?: number;
}

export class MarvinMemoryClient {
  private baseUrl: string;
  private timeout: number;

  /**
   * Initialize the client
   * 
   * @param baseUrl - Base URL for the API
   * @param timeout - Request timeout in milliseconds
   */
  constructor(baseUrl: string = BASE_URL, timeout: number = DEFAULT_TIMEOUT) {
    this.baseUrl = baseUrl;
    this.timeout = timeout;
  }

  /**
   * List memories with pagination
   * 
   * @param options - Options for listing memories
   * @returns Promise resolving to memories and pagination info
   */
  async listMemories(options: ListMemoriesOptions = {}): Promise<MemoriesResponse> {
    const {
      page = 1,
      limit = 50,
      memoryType = null,
      minAlignment = null,
      tags = null
    } = options;

    const params = new URLSearchParams({
      page: page.toString(),
      limit: limit.toString()
    });

    if (memoryType) {
      params.append("memory_type", memoryType);
    }

    if (minAlignment !== null) {
      params.append("min_alignment", minAlignment.toString());
    }

    if (tags) {
      if (Array.isArray(tags)) {
        params.append("tags", tags.join(","));
      } else {
        params.append("tags", tags);
      }
    }

    return this._makeRequest<MemoriesResponse>("GET", "/memories/", { params });
  }

  /**
   * Search memories by semantic similarity
   * 
   * @param query - Search query
   * @param options - Search options
   * @returns Promise resolving to search results
   */
  async searchMemories(query: string, options: SearchMemoriesOptions = {}): Promise<SearchResponse> {
    const {
      limit = 5,
      memoryType = null,
      minAlignment = null,
      tags = null
    } = options;

    const params = new URLSearchParams({
      query,
      limit: limit.toString()
    });

    if (memoryType) {
      params.append("memory_type", memoryType);
    }

    if (minAlignment !== null) {
      params.append("min_alignment", minAlignment.toString());
    }

    if (tags) {
      if (Array.isArray(tags)) {
        params.append("tags", tags.join(","));
      } else {
        params.append("tags", tags);
      }
    }

    return this._makeRequest<SearchResponse>("GET", "/memories/search", { params });
  }

  /**
   * Create a new memory
   * 
   * @param memory - Memory to create
   * @returns Promise resolving to created memory ID
   */
  async createMemory(memory: CreateMemoryOptions): Promise<CreateMemoryResponse> {
    const { content, type, source, tags = null } = memory;

    const data: any = {
      content,
      type,
      source
    };

    if (tags) {
      data.tags = tags;
    }

    return this._makeRequest<CreateMemoryResponse>("POST", "/memories/", { data });
  }

  /**
   * Delete a memory
   * 
   * @param memoryId - Memory ID
   * @returns Promise resolving to deletion status
   */
  async deleteMemory(memoryId: string): Promise<DeleteMemoryResponse> {
    return this._makeRequest<DeleteMemoryResponse>("DELETE", `/memories/${memoryId}`);
  }

  /**
   * Conduct research using Perplexity API
   * 
   * @param query - Research query
   * @param autoApprove - Whether to automatically approve insights
   * @returns Promise resolving to research results
   */
  async conductResearch(query: string, autoApprove: boolean | null = null): Promise<ResearchResponse> {
    const data: any = {
      query
    };

    if (autoApprove !== null) {
      data.auto_approve = autoApprove;
    }

    return this._makeRequest<ResearchResponse>("POST", "/research/", { data });
  }

  /**
   * Get all pending research
   * 
   * @returns Promise resolving to pending research
   */
  async getPendingResearch(): Promise<PendingResearchResponse> {
    return this._makeRequest<PendingResearchResponse>("GET", "/research/");
  }

  /**
   * Get pending research by ID
   * 
   * @param queryId - Research query ID
   * @returns Promise resolving to pending research
   */
  async getPendingResearchById(queryId: string): Promise<ResearchResponse> {
    return this._makeRequest<ResearchResponse>("GET", `/research/${queryId}`);
  }

  /**
   * Approve selected insights for storage
   * 
   * @param queryId - Research query ID
   * @param insightIndices - Indices of insights to approve
   * @returns Promise resolving to approval status
   */
  async approveInsights(queryId: string, insightIndices: number[]): Promise<ApproveInsightsResponse> {
    const data = {
      insight_indices: insightIndices
    };

    return this._makeRequest<ApproveInsightsResponse>("POST", `/research/${queryId}/approve`, { data });
  }

  /**
   * Reject and delete pending research
   * 
   * @param queryId - Research query ID
   * @returns Promise resolving to rejection status
   */
  async rejectResearch(queryId: string): Promise<{ status: string; error?: string }> {
    return this._makeRequest<{ status: string; error?: string }>("DELETE", `/research/${queryId}`);
  }

  /**
   * Manually trigger tweet processing
   * 
   * @param limit - Maximum number of tweets to process
   * @returns Promise resolving to processing results
   */
  async processTweets(limit: number = 10): Promise<ProcessTweetsResponse> {
    const params = new URLSearchParams({
      limit: limit.toString()
    });

    return this._makeRequest<ProcessTweetsResponse>("POST", "/tweets/process", { params });
  }

  /**
   * Make an HTTP request to the API with retry logic
   * 
   * @param method - HTTP method
   * @param endpoint - API endpoint
   * @param options - Request options
   * @returns Promise resolving to response JSON
   * @private
   */
  private async _makeRequest<T>(
    method: string,
    endpoint: string,
    options: RequestOptions = {}
  ): Promise<T> {
    const {
      params = null,
      data = null,
      retryCount = 3,
      retryDelay = 1000
    } = options;

    const url = new URL(`${this.baseUrl}${endpoint}`);
    
    if (params) {
      url.search = params.toString();
    }

    const fetchOptions: RequestInit = {
      method,
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json'
      }
    };

    if (data) {
      fetchOptions.body = JSON.stringify(data);
    }

    let currentRetry = 0;
    let currentDelay = retryDelay;

    while (true) {
      try {
        // Create AbortController for timeout
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), this.timeout);
        fetchOptions.signal = controller.signal;

        const response = await fetch(url.toString(), fetchOptions);
        clearTimeout(timeoutId);

        // Check if the request was successful
        if (!response.ok) {
          // Handle 502 Bad Gateway with retry
          if (response.status === 502 && currentRetry < retryCount) {
            console.log(`Received 502 Bad Gateway. Retrying in ${currentDelay / 1000} seconds...`);
            await new Promise(resolve => setTimeout(resolve, currentDelay));
            currentRetry++;
            currentDelay *= 2;
            continue;
          }

          // Try to get error details from response
          let errorData: any;
          try {
            errorData = await response.json();
          } catch (e) {
            errorData = { error: response.statusText };
          }

          throw new Error(JSON.stringify({
            status: response.status,
            statusText: response.statusText,
            ...errorData
          }));
        }

        // Parse and return the response JSON
        return await response.json() as T;

      } catch (error) {
        // Handle timeout with retry
        if (error.name === 'AbortError' && currentRetry < retryCount) {
          console.log(`Request timed out. Retrying in ${currentDelay / 1000} seconds...`);
          await new Promise(resolve => setTimeout(resolve, currentDelay));
          currentRetry++;
          currentDelay *= 2;
          continue;
        }

        // Handle other errors or max retries reached
        if (currentRetry >= retryCount) {
          console.log('Max retries reached');
        }

        return { error: error.message || 'Unknown error' } as unknown as T;
      }
    }
  }
}

// Example usage
async function runExamples() {
  const client = new MarvinMemoryClient();
  
  try {
    // Example 1: List memories with pagination
    console.log("\n=== Example 1: List memories with pagination ===");
    const result = await client.listMemories({ page: 1, limit: 10 });
    console.log(`Retrieved ${result.memories?.length || 0} memories`);
    console.log(`Pagination info:`, result.pagination);
    
    // Example 2: Search memories
    console.log("\n=== Example 2: Search memories ===");
    const searchResult = await client.searchMemories("AI consciousness", { limit: 3 });
    console.log(`Found ${searchResult.memories?.length || 0} matching memories`);
    
    // Example 3: Create a memory
    console.log("\n=== Example 3: Create a memory ===");
    const createResult = await client.createMemory({
      content: "AI systems should be designed with ethical considerations in mind.",
      type: "thought",
      source: "philosophical reflection",
      tags: ["ethics", "AI", "philosophy"]
    });
    console.log(`Created memory with ID: ${createResult.id}`);
    
    // Example 4: Conduct research
    console.log("\n=== Example 4: Conduct research ===");
    const researchResult = await client.conductResearch(
      "What are the latest developments in AI ethics?",
      false
    );
    console.log(`Research status: ${researchResult.status}`);
    console.log(`Research ID: ${researchResult.query_id}`);
    
  } catch (error) {
    console.error("Error running examples:", error);
  }
}

// Only run examples if this file is executed directly (not imported)
// Check for Node.js environment
if (typeof process !== 'undefined' && process.versions && process.versions.node) {
  try {
    // We're in Node.js - check if this file is being run directly
    if (require.main === module) {
      runExamples().catch(console.error);
    }
  } catch (e) {
    // Ignore errors in environment detection
    console.error("Error in Node.js environment detection:", e);
  }
}

// Export for module usage
export default MarvinMemoryClient;
